### DISCUSSION TITLE
How do I prevent vcpkg from including secondary dependencies in my builds?

### DISCUSSION THREAD

**Discussion by mas-4** [microsoft/vcpkg#24043] [Q&A] [answered]
I have included in my build a static version of zlib as well as a proprietary third party library IMPLODE that includes a second overload of `crc32`. This causes a linking error 

```
Error	C2733	'crc32': you cannot overload a function with 'extern "C"' linkage
```
which I previously solved by renaming `crc32` in zlib to `crc32_zlib`. Now I have been migrating to vcpkg, however, I have found that whether I include zlib in my `vcpkg.json` manifest or not zlib is included because it is a secondary dependency of my other dependencies. For reference, here is my full `vcpkg.json`:

```
{
  "name": "name",
  "version": "0.0.0.1",
  "builtin-baseline": "89295c9fe22e97ca9e380a6c771ccf2ff09dca95",
  "dependencies": [
    {
      "name": "boost",
      "platform": "windows & x64",
      "version>=":  "1.78.0"
    },
    {
      "name": "cpprestsdk",
      "version>=":  "2.10.18",
      "default-features": true,
      "features": [ "websockets" ],
      "platform": "windows & x64"
    },
    {
      "name": "grpc",
      "platform": "windows & x64",
      "version>=": "1.44.0"
    },
    {
      "name": "nlohmann-json",
      "version>=": "3.10.5",
      "platform":  "windows & x64"
    }
  ],
  "overrides": [
    {
      "name": "boost",
      "version": "1.78.0"
    },
    {
      "name": "cpprestsdk",
      "version": "2.10.18"
    },
    {
      "name": "grpc",
      "version": "1.44.0"
    },
    {
      "name": "nlohmann-json",
      "version": "3.10.5"
    }
  ]
}
```

Because of the fact that zlib is automatically included in this vcpkg setup whether I want it to or now, there is a conflict with my local version of zlib with the renamed `crc32`. Specifically, I get dual specified symbols in my compiled zlib.lib vs vcpkg's `zlib1.dll`. 

I can't rename the IMPLODE copy of crc32, and I can't prevent vcpkg from shoehorning its own copy of zlib into my build. The only solution I'm seeing so far is to just not use vcpkg. Am I missing something? It seems to me extremely problematic that secondary dependencies are included in my build whether I want them or not.

**Comment by autoantwort** [comment, marked as answer]
You can use the overlay port functionality to customize the zlib port to use your custom patch.

**Reply by mas-4** [reply]
But wouldn't that break some of the other library's dependencies on zlib? If they call `crc32` but I've renamed it to `crc32_zlib`, wouldn't they have an undefined function?

**Reply by autoantwort** [reply]
Yeah this can happen. Then you also have to patch the dependencies.
